authority/provisioner/sign_ssh_options.go

```diff
@@ -199,67 +208,92 @@ func (m *sshDefaultExtensionModifier) Modify(cert *ssh.Certificate) error {
-	case 0:
-		return errors.New("ssh certificate type has not been set")
-	default:
-		return errors.Errorf("unknown ssh certificate type %d", cert.CertType)
-	}
+		var backdate uint64
+		if cert.ValidAfter == 0 {
+			backdate = uint64(o.Backdate / time.Second)
+			cert.ValidAfter = uint64(now().Truncate(time.Second).Unix())
+		}
+		if cert.ValidBefore == 0 {
+			cert.ValidBefore = cert.ValidAfter + uint64(d/time.Second)
+		}
+		// Apply backdate safely
+		if cert.ValidAfter > backdate {
+			cert.ValidAfter -= backdate
+		}
+		return nil
+	})
+}
 
-	hasLimit := !m.validBefore.IsZero()
+// sshLimitDuration adjusts the duration to min(default, remaining provisioning
+// credential duration). E.g. if the default is 12hrs but the remaining validity
+// of the provisioning credential is only 4hrs, this option will set the value
+// to 4hrs (the min of the two values). It will fail if a CertType has not been
+// set or is not valid.
+type sshLimitDuration struct {
+	*Claimer
+	NotAfter time.Time
+}
 
-	n := now()
-	if cert.ValidAfter == 0 {
-		cert.ValidAfter = uint64(n.Truncate(time.Second).Unix())
-	}
-	certValidAfter := time.Unix(int64(cert.ValidAfter), 0)
-	if hasLimit && certValidAfter.After(m.validBefore) {
-		return errors.Errorf("provisioning credential expiration (%s) is before "+
-			"requested certificate validAfter (%s)", m.validBefore, certValidAfter)
+func (m *sshLimitDuration) Option(o SSHOptions) SSHCertificateModifier {
+	if m.NotAfter.IsZero() {
+		defaultDuration := &sshDefaultDuration{m.Claimer}
+		return defaultDuration.Option(o)
 	}
 
-	if cert.ValidBefore == 0 {
-		certValidBefore := certValidAfter.Add(d)
-		if hasLimit && m.validBefore.Before(certValidBefore) {
-			certValidBefore = m.validBefore
+	return sshModifierFunc(func(cert *ssh.Certificate) error {
+		d, err := m.DefaultSSHCertDuration(cert.CertType)
+		if err != nil {
+			return err
 		}
-		cert.ValidBefore = uint64(certValidBefore.Unix())
-	} else if hasLimit {
-		certValidBefore := time.Unix(int64(cert.ValidBefore), 0)
-		if m.validBefore.Before(certValidBefore) {
-			return errors.Errorf("provisioning credential expiration (%s) is before "+
-				"requested certificate validBefore (%s)", m.validBefore, certValidBefore)
+
+		var backdate uint64
+		if cert.ValidAfter == 0 {
+			backdate = uint64(o.Backdate / time.Second)
+			cert.ValidAfter = uint64(now().Truncate(time.Second).Unix())
 		}
-	}
 
-	return nil
-}
+		certValidAfter := time.Unix(int64(cert.ValidAfter), 0)
+		if certValidAfter.After(m.NotAfter) {
+			return errors.Errorf("provisioning credential expiration (%s) is before requested certificate validAfter (%s)",
+				m.NotAfter, certValidAfter)
+		}
 
-func sshDefaultValidityModifier(c *Claimer) SSHCertificateModifier {
-	return &sshValidityModifier{c, time.Time{}}
-}
+		if cert.ValidBefore == 0 {
+			certValidBefore := certValidAfter.Add(d)
+			if m.NotAfter.Before(certValidBefore) {
+				certValidBefore = m.NotAfter
+				println(2, certValidBefore.String())
```

Comment:
Do we want this println?

Reply:
No, fixed with 77af30bfa395af5a025bb045315359c6fad7977c